### PR TITLE
cli: add --successCache option for repo lint

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

This adds a new caching strategy to the Backstage CLI where we use a very coarse cache to simply store whether the previous run was successful, which then lets us completely skip verification of packages that haven't changed. This is intended to be used in CI, with or without the `--since` flag, although I'm expecting this to same the most time in main branch builds that don't use `--since`. My intention is to add this for tests as well, but starting out with `repo lint` since that's a bit simpler.

The primary components of the cache key for each package is the dependency tree which we grab from `yarn.lock`, the configuration used, as well as the file contents in the package. For the `repo lint` command specifically we can use `ESLint` itself to be a bit more precise, but I consider this a fairly minor optimization and might roll that back if we end up with a simpler shared utility for the cache key generation.

The intention of this cache strategy is only to skip work, never to replay it. The cache is therefore a dead simple list of hashes, and doesn't need to keep track of past output.

By default the cache is stored in `node_modules/.cache/backstage-cli/`, but it can be configured by passing a directory path to the `--cache` option.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/lib/new/factories/backendModule.test.ts [hunk 1]
```diff
@@ -90,7 +90,7 @@ describe('backendModule factory', () => {
       `availability  plugins${sep}test-backend-module-tester-two`,
       'creating      temp dir',
       'Executing Template:',
-      'copying       .eslintrc.js',
+      'templating    .eslintrc.js.hbs',
```
Comment (freben) [inline packages/cli/src/lib/new/factories/backendModule.test.ts:93]: this was an unrelated change really, right?

File: packages/cli/src/commands/repo/lint.ts [hunk 2]
```diff
@@ -30,9 +36,44 @@ function depCount(pkg: BackstagePackageJson) {
   return deps + devDeps;
 }
 
+const CACHE_FILE_NAME = 'lint-cache.json';
+
+type Cache = string[];
```
Comment (freben) [inline packages/cli/src/commands/repo/lint.ts:41]: not worth worrying too much about, but it'd help understanding the on-disk format if it were like

```json
{
  "version": 1,
  "successfullyLintedTreeHashes": [...]
```

or whatnot.

**Review by freben — APPROVED** [review]

**Comment by Rugvip** [comment]
Bonus update, I switched it around to `--successCache` and `--successCacheDir <path>` for two reasons: The `--cache` flag is very generic an likely to conflict with existing build flags. Ideally we're able to keep the flag the same across all `repo` commands that support it, and I figured `--successCache` is unique enough. Couple of other options are `--repoCache`, `--backstageCache`, or `--resultCache`. The `--successCacheDir <path>` split was because the optional `[path]` arg was really confusing and you need to take care to order options properly, and it's not particularly obvious when you get it wrong.

**Comment by Rugvip** [comment]
Gonna ignore the MySQL failure and :shipit:
